funcy-stubs/colls.pyi

```diff
@@ -0,0 +1,342 @@
+from abc import abstractmethod
+from collections.abc import (
+    Container,
+    ItemsView,
+    Iterable,
+    Iterator,
+    KeysView,
+    Mapping,
+    ValuesView,
+)
+from typing import (
+    Any,
+    Callable,
+    Hashable,
+    MutableMapping,
+    Protocol,
+    SupportsIndex,
+    TypeVar,
+    overload,
+)
+
+from typing_extensions import TypeGuard
+
+from ._types import KT, T1, T2, VT, B, Boolean, H, S, T, T_co, T_contra
+
+_Coll = TypeVar("_Coll", bound=Iterable)
+_KT1 = TypeVar("_KT1", bound=Hashable)
+
+class ItemsProtocol(Protocol[T_co]):
+    @abstractmethod
+    def items(self) -> T_co: ...
+
+class ValuesProtocol(Protocol[T_co]):
+    @abstractmethod
+    def values(self) -> T_co: ...
+
+class GetCollectionProtocol(Protocol[T_co]):
+    @abstractmethod
+    def __getitem__(self, i: SupportsIndex, /) -> T_co: ...
+
+class SetCollectionProtocol(Protocol[T_contra]):
+    @abstractmethod
+    def __setitem__(self, key: SupportsIndex, value: T_contra, /) -> None: ...
+    @abstractmethod
+    def __getitem__(self, i: SupportsIndex, /) -> Any: ...
+
+class DelCollectionProtocol(Protocol):
+    @abstractmethod
+    def __delitem__(self, key: SupportsIndex, /) -> None: ...
+    @abstractmethod
+    def __getitem__(self, i: SupportsIndex, /) -> Any: ...
+
+_DelCollectionType = TypeVar("_DelCollectionType", bound=DelCollectionProtocol)
+
+@overload
+def empty(coll: Iterator[T]) -> Iterator[T]: ...
+@overload
+def empty(coll: KeysView[KT]) -> list[KT]: ...
+@overload
+def empty(coll: ValuesView[VT]) -> list[VT]: ...
+@overload
+def empty(coll: ItemsView[KT, VT]) -> list[tuple[KT, VT]]: ...
+@overload
+def empty(coll: _Coll) -> _Coll: ...
+@overload
+def iteritems(coll: ItemsProtocol[T]) -> T: ...
+@overload
+def iteritems(coll: T) -> T: ...
+@overload
+def itervalues(coll: ValuesProtocol[T]) -> T: ...
+@overload
+def itervalues(coll: T) -> T: ...
+@overload
+def join(colls: Iterable[str | bytes]) -> str: ...
+@overload
+def join(colls: Iterable[MutableMapping[KT, VT]]) -> MutableMapping[KT, VT]: ...
+@overload
+def join(colls: Iterable[set[H]]) -> set[H]: ...
+@overload
+def join(colls: Iterable[Iterable[T]]) -> Iterable[T]: ...
+@overload
+def join(colls: Iterable[range]) -> Iterator[int]: ...
```

Comment:
Тут pyright подсказывает, что пятый overload перекрывается четвертым:
```bash
Overload 5 for "join" will never be used because its parameters overlap overload 4
```
Если поменять местами четвертый с пятым ошибка уходит. Что думаешь, чтобы их поменять местами/убрать пятый или же это ошибка чекера?

Reply:
Mypy выдает аналогичную ошибку:
```bash
Overloaded function signature 5 will never be matched: signature 4's parameter type(s) are the same or broader
```

Reply:
Исправил + добавил pre-commit с pyright и mypy